Feat : 댓글 API 기능 구현

gombasan opened the pull request:
# 작업사항
- #8 
Comment API 를 추가하였습니다.

- 댓글 등록과 수정 삭제를 추가하였습니다.
- 대댓글을 등록 할 수 있도록 추가하였습니다.
- 댓글 조회는 post 조회 시 함께 조회 되므로 Comment API 에서는 등록,수정,삭제만 하도록 하였습니다.
- Comment 의 속성 중 commentId - > parentId 로 변경하였습니다. 이유는 댓글이 대댓글이 생기며 계층이 생기므로 commentId 는 곧 부모의 Id 라고 생각하여 수정했습니다.
- MapStruct 를 사용하여 dto 를 매핑하도록 추가하였습니다.

# 특이사항
- Post 상세 조회 시 Member 를 조회 하는 쿼리문이 하나 더 생성됩니다...!
**- `joinCommentAndLikeFindById` 이 메서드를 실행할 때 어째서인지.. 멤버를 조회하는 쿼리를 하나 더 보내네요...**

gombasan committed:
feat(Comment) : Comment 등록 ,수정 추가

- 댓글 등록과 수정을 기능을 추가하였습니다.

gombasan committed:
feat(Comment) : CommentRepository QueryDSL 추가

- 기존 게시글과 같이 Spring-Data-JPA 와 QueryDSL 을 같이 사용하기 위해 CustomRepository 를 상속받아 사용합니다.

gombasan committed:
chore(Comment) : CommentController 반환 타입 변경

- 댓글 생성 및 수정 시 body 에 CommentResponse 를 반환하도록 수정하였습니다.
- CommentResponse 에 정적 팩토리 메서드를 오버라이딩하여 List 형태의 Comment 외의 Comment 를 매개변수로 받도록 생성하였습니다.
- 대댓글 기능 구현을 위해 URL 매핑과 메서드명을 정의했습니다.
- Comment 계층형 쿼리를 사용하여 대댓글을 사용하기 위해 commentId 를 parentId 로 변경하였습니다.

gombasan committed:
chore(Comment) : Comment 생성자 수정

- commentId -> parentId 로 수정하였습니다.

gombasan committed:
feat(Comment) : 대댓글 기능 추가

- 대댓글 추가, 삭제를 추가하였습니다.
- URL 이 comments 보다 comment 가 적절하다고 생각하여 수정하였습니다.
- Member 와 Post 를 fetchJoin 으로 즉시로딩하도록 설정하였습니다.

gombasan committed:
feat(Post) : 게시글 조회 시 대댓글 내용 추가

- 게시글 상세 조회 할 경우 대댓글은 제외하고 가져오도록 서브쿼리를 추가하였습니다.
- Comment 의 parentId 가 있을 경우 조회하지 않도록 추가하였습니다.
- 대댓글은 따로 API 호출로 조회 할 예정이므로, 게시글 조회시에는 제외 하도록 하였습니다.
- 게시글 전체 조회 시에는 대댓글의 갯수도 포함하여 반환합니다.

gombasan committed:
chore(Post) : 게시글 상세조회 수정

- 서브쿼리에 불필요한 where 문 삭제하였습니다.

gombasan committed:
chore : Comment 속성 변경

- comment_id -> parent_id 로 변경하였습니다.

JoosungKwon commented on src/main/java/com/clover/habbittracker/domain/comment/api/CommentController.java:
```diff
@@ -1,16 +1,74 @@
 @RestController
 @RequiredArgsConstructor
-@RequestMapping("/comments")
+@RequestMapping("/posts")
```
`"/{postId}"`도 전부 중복되는 것을 보니 포함해도 괜찮을 것 같네요

gombasan replied:
휴...다행이네요..ㅋㅋㅋㅋㅋ
comments 를 사용할려고 하다보니 url 이 이상해지는거같아서 바꿔버렸습니다!

JoosungKwon commented on src/main/java/com/clover/habbittracker/domain/comment/api/CommentController.java:
```diff
@@ -1,16 +1,74 @@
+		@AuthenticationPrincipal Long memberId,
+		@PathVariable Long postId,
+		@RequestBody CommentRequest request
+	) {
```
보기 훨씬 좋네요~👍

gombasan replied:
이제는 습관처럼 매개변수 하나 쓰고 엔터를 칩니다 ㅎㅎ

JoosungKwon replied:
저는 컨트롤러에서만 그렇게 하긴 합니다 ㅋㅋ

JoosungKwon commented on src/main/java/com/clover/habbittracker/domain/comment/api/CommentController.java:
```diff
@@ -1,16 +1,74 @@
+	}
+
+	@GetMapping("/{postId}/comment/{commentId}/reply")
+	public ResponseEntity<ApiResponse<List<CommentResponse>>> getReplyList(
```
LGTM 👍

gombasan replied:
너무 길어지는게 아닌가 싶었는데, 감사합니다~~

JoosungKwon commented on src/main/java/com/clover/habbittracker/domain/comment/api/CommentController.java:
```diff
@@ -1,16 +1,74 @@
+		@RequestBody CommentRequest request
+	) {
+		commentService.createReply(memberId, commentId, postId, request);
+		return ResponseEntity.ok().body(ApiResponse.success());
```
이 경우에는 201을 보내는게 맞지 않을까요? 어떻게 생각하시나요?

gombasan replied:
생성은 다른 얘들도 상태코드가 201 이기 떄문에 201로 통일 해야댈거같아요!

[ApiResponse 와 ResponseEntity Comment](https://github.com/Clover-Habiters/backend/pull/17#discussion_r1257476494)

위 코맨트와 합쳐서 아래와 같이 새로 수정해봤습니다~!
[fix : CommentController 반환 타입 수정](https://github.com/Clover-Habiters/backend/pull/17/commits/a8135c923f09dcbb92215b9d066c701b9b55103b)

JoosungKwon replied:
오호 Status를 적용해서 ResponseEntity를 빼셨군요! 해당 방법도 괜찮아보이네요!
다른 동료가 예전 코드 리뷰 받은 내용이 있어서 한번 확인해보시면 좋을 것 같아요~~ 

- [참고] : https://github.com/mamission/jwp_cat_picture_search/pull/2#discussion_r1159346139

JoosungKwon commented on src/main/java/com/clover/habbittracker/domain/comment/api/CommentController.java:
```diff
@@ -1,16 +1,74 @@
+	}
+
+	@PutMapping("/{postId}/comment/{commentId}")
+	public ResponseEntity<ApiResponse<CommentResponse>> updateComment(
```
`ApiResponse`하고 `ResponseEntity`가 같이 가는 것 같은데 합쳐볼 순 없을까요?

JoosungKwon commented on src/main/java/com/clover/habbittracker/domain/comment/dto/CommentResponse.java:
```diff
@@ -11,4 +11,8 @@ public record CommentResponse(
 		return comments.stream().map(comment -> new CommentResponse(comment.getId(), comment.getContent())).toList();
 	}
+
+	public static CommentResponse from(Comment comment) {
```
저희 이펙티브 자바할때 정적팩터리 메서드 명명 규칙 확인하셨나요?

gombasan replied:
from : 매개변수 하나를 받아서 해당 타입의 인스턴스를 반환하는 메서드
of : 여러 매개변수를 받아 적합한 타입의 인스턴스를 반환하는 집계 메서드

이 두개를 고민했었는데, 집계 메서드가 아니고 매개변수도 하나라서 from 으로 작성했었습니다!

JoosungKwon commented on src/main/java/com/clover/habbittracker/domain/comment/entity/Comment.java:
```diff
@@ -50,14 +51,18 @@ public class Comment {
+	}
+
+	public void updateComment(CommentRequest request) {
+		this.content = request.content();
```
유효성 검증이 필요할 것 같아요!

gombasan replied:
주성님이 만들어주신 좋은 ValidateUtils 를 사용해서 유효성 검증을 추가하였습니다~!

[feat : Comment 유효성 검증 추가](https://github.com/Clover-Habiters/backend/pull/17/commits/9a4ad20b8761e3f4d8ea613ce9db3bdc900c31c7)

JoosungKwon commented on src/main/java/com/clover/habbittracker/domain/comment/repository/CommentCustomRepositoryImpl.java:
```diff
@@ -0,0 +1,45 @@
+@RequiredArgsConstructor
+public class CommentCustomRepositoryImpl implements CommentCustomRepository{
+
+	private final JPAQueryFactory jpaQueryFactory;
+	@Override
```
형님 코드 재정렬 자동화 적용하셔야 할듯! 저도 요새 인텔리제이가 자꾸 지멋대로 바꿔서 조만간 손 한번 보겠음당 ㅋㅋ

gombasan replied:
오... 저는 그 인텔리제이 `commend` + `options` + `l`  로 사용해서 재정렬 하고있었는데, 코딩 컨벤션 적용할까요?!

JoosungKwon replied:
이부분은 따로 이야기한번 나눠봅시닷~

JoosungKwon commented on src/main/java/com/clover/habbittracker/domain/comment/service/CommentService.java:
```diff
@@ -1,14 +1,19 @@
+import com.clover.habbittracker.domain.comment.dto.CommentResponse;
 
-import lombok.RequiredArgsConstructor;
+public interface CommentService {
```
Service에 인터페이스가 필요하다고 생각하시나요?

gombasan replied:
아주 좋은 질문입니다!

인터페이스를 쓰는 이유가 OCP 를 지키기 위함인데, 요즘 들어 Service 구현체가 바뀔일이 있을까 싶네요.
일단은 역할을 분리해서 Controller 가 Service 의 역할만을 의존하도록 하기 위해 인터페이스로 분리했습니다!

JoosungKwon replied:
저도 Service는 바뀔일이 없을 것 같다고 생각해서.. 일단은 적용하셨으니 추후 리팩터링에서 이야기해보죠!

JoosungKwon commented on src/main/java/com/clover/habbittracker/domain/comment/service/CommentServiceImpl.java:
```diff
@@ -0,0 +1,94 @@
+
+	private void verifyPermissions(Member member, Long memberId) {
+		if (!member.getId().equals(memberId)) {
+			throw new PermissionDeniedException(memberId);
```
```suggestion
		if (!Objects.equals(member.getId(), memberId)) {
			throw new PermissionDeniedException(memberId);
```

이런식으로 하면 NPE를 방지할 수 있습니다!

gombasan replied:
memberId 가 Null 이 들어올거라는걸 생각도 못했네요!
감사합니다 해당 내용 수정했습니다!
[feat : CommentMapper 추가](https://github.com/Clover-Habiters/backend/pull/17/commits/c83221d1461d9197f52d866c4633fac9fc04cebd)

JoosungKwon replied:
여러 사람이 같이 코드를 작성하게 되면 언제 어떤 상황이 발생할 것이다 장담할 수 없게 됩니다! 그렇기 때문에 위험 요소는 최대한 배제하는 것이 좋습니다!

JoosungKwon commented on src/main/java/com/clover/habbittracker/domain/comment/service/CommentServiceImpl.java:
```diff
@@ -0,0 +1,94 @@
+		Comment comment = Comment.builder()
+			.member(member)
+			.post(post)
+			.parentId(commentId)
```
commentId도 있는지 확인해 볼 필요가 있을까요? 이 부분은 어떻게 생각하시나요?

gombasan replied:
확인해 볼 필요가 있을거같아요!

현재 구성으로는 Comment 의 Id 를 Comment 가 parentId 라는 속성으로 사용합니다!
따라서 코드에서 commentId 가 실제로 존재하지 않는 Id 라면 댓글을 등록해도 조회가 안될것같아요!

아니면 구성을 바꾸는 방법으로 해결해야할거같습니다!

JoosungKwon replied:
좋습니다! 저도 부모 댓글이 있는지 확인하는 동작까지 있는 것이 나쁘지 않다고 생각이 드네요!

JoosungKwon reviewed (commented):
@gombasan  고생하셨습니다~!👍
일부러 리뷰어를 안달아 두셨군요 담에 다시 달아두시면 재리뷰하겠습니다~~!! 😄

gombasan committed:
feat : Comment 유효성 검증 추가

- 댓글 수정 시 유효성 검증을 추가하였습니다.
- 만들어진 ValidateUtils 클래스의 메서드를 활용하였습니다.

gombasan committed:
fix : CommentController 반환 타입 수정

- ResponseEntity 와 ApiResponse 를 같이 사용함에 있어 반환타입이 장황해졌습니다.
- ResponseEntity 를 반환하는 이유는 상태코드를 같이 전달하는 용도로 사용하므로 어노테이션으로 변경하였습니다.
- 상태코드를 정하지 않는다면 기본이 200 OK 를 반환하므로 200 인 경우 어노테이션을 제거하였습니다.

gombasan committed:
feat : CommentMapper 추가

- CommentMapper 를 추가하였습니다.
- 기존 정적 팩터리 메서드를 제거하고, CommentMapper 를 사용하도록 변경하였습니다.
- 사용자 수정,삭제 권한 확인시 NPE 가 발생 할 수 있어 Objects 유틸 클래스를 사용하여 처리하도록 변경하였습니다.

gombasan committed:
feat : CommentServiceTest 추가

- CommentService 테스트 코드를 추가하였습니다.

gombasan committed:
feat : CommentRepositoryTest 추가

- CommentRepository 테스트 코드를 추가하였습니다.

JoosungKwon commented on src/main/java/com/clover/habbittracker/domain/comment/mapper/CommentMapper.java:
```diff
@@ -0,0 +1,34 @@
+import com.clover.habbittracker.domain.post.entity.Post;
+
+@Mapper(componentModel = "spring", unmappedTargetPolicy = IGNORE)
+public interface CommentMapper {
```
알아서 잘 사용하시네요~~~ 👍

gombasan replied:
진짜 써보니까 너무 편해서 버릴수가 없습니다...허허허
테스트 코드도 작성해서 올리도록 하겠습니다!

JoosungKwon commented on src/test/java/com/clover/habbittracker/domain/comment/repository/CommentRepositoryTest.java:
```diff
@@ -0,0 +1,116 @@
+
+@DataJpaTest
+@Import(JpaConfig.class)
+public class CommentRepositoryTest {
```
RepositoryTest 하면서 쿼리 직접 확인하시나요~? Service 테스트도 그렇고 예상치 못한 쿼리는 없는지 확인해보면 좋을 것 같네요~!

gombasan replied:
생각해보니 따로 확인을 안했습니다...!
피드백 감사합니다!

JoosungKwon reviewed (commented):
@gombasan 
이왕 테스트 추가하신 김에 Restdocs 적용하고 커버리지 맞춘 후에 Merge하면 될 것 같아요~~

gombasan commented:
@JoosungKwon 
리뷰 감사합니다~! ㅎㅎ
대댓글 쉽게 생각했었는데 쉽지않네요....
테스트 코드 마저 작성하도록 하겠습니당 😅😅

JoosungKwon commented on src/main/java/com/clover/habbittracker/domain/comment/api/CommentController.java:
```diff
@@ -1,16 +1,73 @@
 	private final CommentService commentService;
+
+	@PostMapping("/{postId}/comment")
+	@ResponseStatus(CREATED)
+	public ApiResponse<CommentResponse> createComment(
```
@gombasan 
이거 이야기를 깜빡했네요! 저는 ResponseStatus를 주로 반환이 없는 경우에만 사용합니다! 한번 참고 해보시면 좋을 것 같아요(https://joojimin.tistory.com/54)

JoosungKwon commented:
@gombasan 형님 일단 머지 하시지요 ㅋㅋ 
하고 리팩터링하면서 맞추면 될것 같아요!

JoosungKwon committed:
Merge branch 'develop' into feat/8-add-comment

gombasan committed:
feat(Test) : PostProvider 추가

- 커뮤니티 내에서 테스트를 할 때 게시글이 항상 필요하므로 Util 클래스를 생성하여 관리하고자 추가했습니다.
- Member 를 매개변수로 받아 게시글을 생성합니다.

gombasan committed:
chore(Test) : Post 생성 코드 수정

- PostProvider 가 추가됨에 따라 PostBuilder 를 제거하였습니다.

gombasan committed:
fix(Test) : MemberProvider 오류 수정

- ID 를 직접 넣지 않도록 변경하였습니다.

gombasan committed:
feat(Test) : CommentTest 추가

- Comment 생성과 수정 및 validCheck 테스트 코드를 추가하였습니다.

gombasan committed:
feat(Test) : CommentControllerTest 추가

- 기존 Controller 테스트와 동일하게 mockMvc 를 활용하여 테스트를 진행했습니다.
- RestDocs 내용도 함께 추가하였습니다.

gombasan committed:
feat(Test) : CommentMapperTest 추가

- SpringTest 없이 테스트를 진행하기 위해 구현체를 직접 의존하도록 설정하였습니다.
- mapping 된 결과가 정상적인지 확인하는 목적을 위해 작성하였습니다.

gombasan committed:
Merge remote-tracking branch 'origin/feat/8-add-comment' into feat/8-add-comment

gombasan commented:
@JoosungKwon 
테스트 코드 일단 작성은 했습니다~!
머지하도록 할게요!
